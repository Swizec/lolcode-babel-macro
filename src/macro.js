import { createMacro } from "babel-plugin-macros";
import { parser } from "@swizec/loljs";

import JSify from "./JSify";
import stdlib from "./lolstdlib";

export default createMacro(myMacro);

function myMacro({ references, state, babel }) {
    references.lolcode.forEach(referencePath => {
        const compiled = compileLolcode(referencePath);

        referencePath.parentPath.replaceWithSourceString(compiled);
    });
}

function compileLolcode(referencePath) {
    const source = referencePath.parentPath.node.quasi.quasis
        .map(node => node.value.raw)
        .join("");

    const ast = parser.parse(source);
    const jsify = new JSify();

    return `(function (stdlib) {
       ${jsify.compile(ast)} 
    })(${stdlib})`;
}
