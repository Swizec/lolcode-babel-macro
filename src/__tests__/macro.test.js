const path = require("path");
const pluginTester = require("babel-plugin-tester");
const plugin = require("babel-plugin-macros");
const prettier = require("prettier");

pluginTester({
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename
    },
    formatResult(result) {
        return prettier.format(result, {
            trailingComma: "es5"
        });
    },
    tests: {
        "hello world": {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    VISIBLE "HAI WORLD!"
                    KTHXBYE
                \`;
            `
        },
        assignment: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    I HAS A VAR
                    VAR R 3
                    KTHXBYE
                \`;
            `
        },
        declaration: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    I HAS A VAR
                    KTHXBYE
                \`;
            `
        },
        functioncall: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    HOW DUZ I IZCOLOR YR COLOR
                        VISIBLE COLOR
                    IF U SAY SO
                    IZCOLOR "yellow" MKAY
                    KTHXBYE
                \`;
            `
        },
        "functiondefinition with args": {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    HOW DUZ I IZCOLOR YR COLOR
                        VISIBLE color
                    IF U SAY SO
                    KTHXBYE
                \`;
            `
        },
        "functiondefinition with return": {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    HOW DUZ I IZCOLOR YR COLOR
                        FOUND YR color
                    IF U SAY SO
                    KTHXBYE
                \`;
            `
        },
        functiondefinition: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                     HOW DUZ I IZYELLOW
                        VISIBLE "yellow"
                    IF U SAY SO
                    KTHXBYE
                \`;
            `
        },
        gimmeh: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        if: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    WIN, O RLY?
                        YA RLY
                            VISIBLE WIN
                        NO WAI
                            VISIBLE FAIL
                    OIC
                    KTHXBYE
                \`;
            `
        },
        indexer: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        loop: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        loopcondition: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        noop: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        return: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        },
        switch: {
            error: false,
            code: `
                import { lolcode } from '../macro';
                const code = lolcode\`
                    HAI
                    
                    KTHXBYE
                \`;
            `
        }
    }
});
