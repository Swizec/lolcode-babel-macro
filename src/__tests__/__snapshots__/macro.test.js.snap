// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros assignment: assignment 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    I HAS A VAR
    VAR R 3
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  let VAR = null;
  VAR = 3;
};

"
`;

exports[`macros declaration: declaration 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    I HAS A VAR
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  let VAR = null;
};

"
`;

exports[`macros functioncall: functioncall 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    HOW DUZ I IZCOLOR YR COLOR
        VISIBLE COLOR
    IF U SAY SO
    IZCOLOR \\"yellow\\" MKAY
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  function IZCOLOR(color) {
    console.log(color);
  }

  IZCOLOR(\\"yellow\\");
};

"
`;

exports[`macros functiondefinition with args: functiondefinition with args 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    HOW DUZ I IZCOLOR YR COLOR
        VISIBLE color
    IF U SAY SO
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  function IZCOLOR(color) {
    console.log(color);
  }
};

"
`;

exports[`macros functiondefinition with return: functiondefinition with return 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    HOW DUZ I IZCOLOR YR COLOR
        FOUND YR color
    IF U SAY SO
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  function IZCOLOR(color) {
    return color;
  }
};

"
`;

exports[`macros functiondefinition: functiondefinition 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
     HOW DUZ I IZYELLOW
        VISIBLE \\"yellow\\"
    IF U SAY SO
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  function IZYELLOW() {
    console.log(\\"yellow\\");
  }
};

"
`;

exports[`macros gimmeh: gimmeh 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros hello world: hello world 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    VISIBLE \\"HAI WORLD!\\"
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  console.log(\\"HAI WORLD!\\");
};

"
`;

exports[`macros if: if 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    WIN, O RLY?
        YA RLY
            VISIBLE WIN
        NO WAI
            VISIBLE FAIL
    OIC
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {
  true;

  if (undefined) {
    console.log(true);
  } else {
    console.log(false);
  }
};

"
`;

exports[`macros indexer: indexer 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros loop: loop 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros loopcondition: loopcondition 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros noop: noop 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros return: return 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;

exports[`macros switch: switch 1`] = `
"
import { lolcode } from '../macro';
const code = lolcode\`
    HAI
    
    KTHXBYE
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

const code = function() {};

"
`;
